### PR TITLE
Add a non-sticky network check of metadata server detection to google default credentials

### PR DESCRIPTION
**Pull request by yihuazhang**
This PR brings back original behavior of metadata server detection (i.e., doing network test), but makes it non-sticky if the detection result is negative. It will provide a fix to use cases that cannot provide platform hints or set grpc-provided environment variables. 

There exists an inconsistency in implementing the network test across different languages: 

C++: 1 sec deadline + no retries (original behavior). 
Go: 5 sec deadline + no retries. 
Python: 3 sec deadline + no retries.
Java: 0.5 sec deadline + 3 retries.

Since the connection to metadata server will be local in most cases, 1 sec should be sufficient IMO. Please let me know if you think otherwise. 

Original PR that removes the network check: #15408

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,718      Total (=)      2,019,718

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,625      Total (>)     11,154,122

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,718      Total (=)      2,019,718

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,628      Total (>)     11,154,123

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by yihuazhang** [comment]
@stevewolter gRPC does not allow me to set you as a reviewer, but feel free to leave comments.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                      FILE SIZE
 ++++++++++++++ GROWING                                                                        ++++++++++++++
  +0.0%    +160 [None]                                                                         +9.25Ki  +0.1%
      +0.0%    +160 [Unmapped]                                                                     +9.25Ki  +0.1%
      [NEW]      +8 g_polling_mu                                                                         0  [ = ]
      [NEW]      +4 g_is_on_gce                                                                          0  [ = ]
      [NEW]      +4 g_metadata_server_available                                                          0  [ = ]
      [NEW]      +4 g_metadata_server_detection_done                                                     0  [ = ]
   +40% +1.06Ki src/core/lib/security/credentials/google_default/google_default_credentials.cc +1.06Ki   +40%
       +83%    +832 grpc_google_default_credentials_create                                            +832   +83%
      [NEW]    +218 on_metadata_server_detection_http_response                                        +218  [NEW]
      +8.8%     +19 google_default_create_security_connector                                           +19  +8.8%
       +29%     +14 [Unmapped]                                                                         +14   +29%
      [NEW]      +5 destroy_pollset                                                                     +5  [NEW]

  +0.1% +1.22Ki TOTAL                                                                          +10.3Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc [hunk 1]
```diff
@@ -89,15 +91,20 @@ static grpc_security_status google_default_create_security_connector(
   bool use_alts =
       is_grpclb_load_balancer || is_backend_from_grpclb_load_balancer;
   grpc_security_status status = GRPC_SECURITY_ERROR;
+  /* Return failure if ALTS is selected but not running on GCE. */
+  if (use_alts && !g_is_on_gce) {
+    goto end;
```
Comment (jiangtaoli2016) [inline src/core/lib/security/credentials/google_default/google_default_credentials.cc]: We need to log ERROR here.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc [hunk 2]
```diff
@@ -49,9 +49,11 @@
 
 /* -- Default credentials. -- */
 
-static int g_compute_engine_detection_done = 0;
-static int g_need_compute_engine_creds = 0;
+static int g_metadata_server_detection_done = 0;
+static int g_metadata_server_available = 0;
+static int g_is_on_gce = 0;
 static gpr_mu g_state_mu;
+static gpr_mu* g_polling_mu;
```
Comment (stevewolter) [inline src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Could you please document what the mutex is protecting? IIUC, it's the pollset and is_done, but I'm not entirely sure.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc [hunk 3]
```diff
@@ -49,9 +49,11 @@
 
 /* -- Default credentials. -- */
 
-static int g_compute_engine_detection_done = 0;
-static int g_need_compute_engine_creds = 0;
+static int g_metadata_server_detection_done = 0;
```
Comment (stevewolter) [inline src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Could you please add a comment explaining why this variable is sticky for positive results, but not negative? It's totally the right thing to do, but I'm sure posterity would love to know.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc [hunk 4]
```diff
@@ -202,16 +295,25 @@ grpc_channel_credentials* grpc_google_default_credentials_create(void) {
...
-  /* At last try to see if we're on compute engine (do the detection only once
-     since it requires a network test). */
-  if (!g_compute_engine_detection_done) {
-    g_need_compute_engine_creds = g_gce_tenancy_checker();
-    g_compute_engine_detection_done = 1;
+
+  /* Try a platform-provided hint for GCE. */
+  if (!g_metadata_server_detection_done) {
+    g_is_on_gce = g_gce_tenancy_checker();
+    g_metadata_server_detection_done = g_is_on_gce;
```
Comment (stevewolter) [inline src/core/lib/security/credentials/google_default/google_default_credentials.cc]: The two g_metadata_server variables are redundant: Whenever detection_done is true, available is true, and they only differ after flush(), at which time we'll overwrite available before the next read. Should we keep only g_metadata_server_available?

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc (hunk 1 above)
Comment (yihuazhang) [reply src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Sure, the ERROR log is added.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc (hunk 2 above)
Comment (yihuazhang) [reply src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Yes, you are right. It is protecting `pollent` and `is_done` bit. I added a comment to make it more clear.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc (hunk 3 above)
Comment (yihuazhang) [reply src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Good suggestion. A comment is added. PTAL.

File: src/core/lib/security/credentials/google_default/google_default_credentials.cc (hunk 4 above)
Comment (yihuazhang) [reply src/core/lib/security/credentials/google_default/google_default_credentials.cc]: Good catch. I removed the redundancy by using `g_metadata_server_available` only.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                      FILE SIZE
 ++++++++++++++ GROWING                                                                        ++++++++++++++
  +0.0%    +240 [None]                                                                         +9.08Ki  +0.1%
      +0.0%    +240 [Unmapped]                                                                     +9.08Ki  +0.1%
      [NEW]      +8 g_polling_mu                                                                         0  [ = ]
      [NEW]      +4 g_is_on_gce                                                                          0  [ = ]
      [NEW]      +4 g_metadata_server_available                                                          0  [ = ]
   +41% +1.08Ki src/core/lib/security/credentials/google_default/google_default_credentials.cc +1.08Ki   +41%
       +80%    +808 grpc_google_default_credentials_create                                            +808   +80%
      [NEW]    +218 on_metadata_server_detection_http_response                                        +218  [NEW]
       +25%     +55 google_default_create_security_connector                                           +55   +25%
       +38%     +18 [Unmapped]                                                                         +18   +38%
      [NEW]      +5 destroy_pollset                                                                     +5  [NEW]

  +0.1% +1.31Ki TOTAL                                                                          +10.2Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,735      Total (=)      2,019,735

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,520      Total (>)     11,154,125

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,751      Total (=)      2,019,751

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,154,514      Total (>)     11,154,114

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by yihuazhang** [comment]
MSAN failure: #17375

**Comment by yihuazhang** [comment]
interop cloud-to-prod failures: #13379

**Pull request merged**